web/src/enterprise/campaigns/close/CampaignCloseChangesetsList.tsx

```diff
@@ -0,0 +1,173 @@
+            queryChangesets({
+                externalState: ChangesetExternalState.OPEN,
+                publicationState: ChangesetPublicationState.PUBLISHED,
+                checkState: null,
+                reviewState: null,
+                first: args.first ?? null,
+                campaign: campaignID,
+                onlyPublishedByThisCampaign: true,
+            }).pipe(repeatWhen(notifier => notifier.pipe(delay(5000)))),
+        [campaignID, queryChangesets]
+    )
+
+    const containerElements = useMemo(() => new Subject<HTMLElement | null>(), [])
+    const nextContainerElement = useMemo(() => containerElements.next.bind(containerElements), [containerElements])
+
+    const hoverOverlayElements = useMemo(() => new Subject<HTMLElement | null>(), [])
+    const nextOverlayElement = useCallback((element: HTMLElement | null): void => hoverOverlayElements.next(element), [
+        hoverOverlayElements,
+    ])
+
+    const closeButtonClicks = useMemo(() => new Subject<MouseEvent>(), [])
+    const nextCloseButtonClick = useCallback((event: MouseEvent): void => closeButtonClicks.next(event), [
+        closeButtonClicks,
+    ])
+
+    const componentRerenders = useMemo(() => new Subject<void>(), [])
+
+    const hoverifier = useMemo(
+        () =>
+            createHoverifier<RepoSpec & RevisionSpec & FileSpec & ResolvedRevisionSpec, HoverMerged, ActionItemAction>({
+                closeButtonClicks,
+                hoverOverlayElements,
+                hoverOverlayRerenders: componentRerenders.pipe(
+                    withLatestFrom(hoverOverlayElements, containerElements),
+                    map(([, hoverOverlayElement, relativeElement]) => ({
+                        hoverOverlayElement,
+                        // The root component element is guaranteed to be rendered after a componentDidUpdate
+                        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                        relativeElement: relativeElement!,
+                    })),
+                    // Can't reposition HoverOverlay if it wasn't rendered
+                    filter(property('hoverOverlayElement', isDefined))
+                ),
+                getHover: hoveredToken =>
+                    getHover(getLSPTextDocumentPositionParameters(hoveredToken), { extensionsController }),
+                getDocumentHighlights: hoveredToken =>
+                    getDocumentHighlights(getLSPTextDocumentPositionParameters(hoveredToken), { extensionsController }),
+                getActions: context => getHoverActions({ extensionsController, platformContext }, context),
+                pinningEnabled: true,
+            }),
+        [
+            closeButtonClicks,
+            containerElements,
+            extensionsController,
+            hoverOverlayElements,
+            platformContext,
+            componentRerenders,
+        ]
+    )
+    useEffect(() => () => hoverifier.unsubscribe(), [hoverifier])
```

Comment:
Can't you do this?

```suggestion
    useEffect(() => hoverifier.unsubscribe(), [hoverifier])
```

Or even this?

```suggestion
    useEffect(hoverifier.unsubscribe, [hoverifier])
```

Reply:
No, because both your suggestions would be called immediately, and hence the hoverifier is disposed.
A function passed to useEffect may return a function that is run when the component is disposed, that's why we return a double arrow fn here

Reply:
So it would need to be this then:

```suggestion
    useEffect(() => hoverifier.unsubscribe, [hoverifier])
```

The function returns `hoveririfer.unsubscribe` which is then called when the component is disposed, right?

(I actually had that as the first suggestion but then was unsure)

Reply:
Yes that works I believe (if it doesn't require `.bind(hoverifier)`. They're equally hard to understand though I guess 😬